api/v1/etcd_types.go

```diff
@@ -17,135 +17,115 @@ limitations under the License.
-	ImageVersion string `json:"imageVersion"`
-	// +optional
-	FullSnapshotSchedule string `json:"fullSnapshotSchedule,omitempty"`
+	ClientTLSSecretRef corev1.SecretReference
+}
+
+// BackupSpec defines parametes associated with the full and delta snapshots of etcd
+type BackupSpec struct {
 	// +optional
-	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+	Port *int `json:"port,omitempty"`
 	// +optional
-	PullPolicy corev1.PullPolicy `json:"pullPolicy,omitempty"`
+	TLS *TLSConfig `json:"tls,omitempty"`
+	// +required
+	Version string `json:"version"`
 	// +optional
-	GarbageCollectionPolicy GarbageCollectionPolicy `json:"garbageCollectionPolicy,omitempty"`
+	Store *StoreSpec `json:"store,omitempty"`
 	// +optional
-	GarbageCollectionPeriod string `json:"garbageCollectionPeriod,omitempty"`
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
 	// +optional
-	EtcdQuotaBytes int `json:"etcdQuotaBytes,omitempty"`
+	FullSnapshotSchedule *string `json:"fullSnapshotSchedule,omitempty"`
 	// +optional
-	EtcdConnectionTimeout string `json:"etcdConnectionTimeout,omitempty"`
+	GarbageCollectionPolicy *GarbageCollectionPolicy `json:"garbageCollectionPolicy,omitempty"`
 	// +optional
-	SnapstoreTempDir string `json:"snapstoreTempDir,omitempty"`
+	GarbageCollectionPeriod *time.Duration `json:"garbageCollectionPeriod,omitempty"`
 	// +optional
-	DeltaSnapshotPeriod string `json:"deltaSnapshotPeriod,omitempty"`
+	DeltaSnapshotPeriod *time.Duration `json:"deltaSnapshotPeriod,omitempty"`
 	// +optional
-	DeltaSnapshotMemoryLimit int `json:"deltaSnapshotMemoryLimit,omitempty"`
+	DeltaSnapshotMemoryLimit *resource.Quantity `json:"deltaSnapshotMemoryLimit,omitempty"`
 }
 
 // EtcdSpec defines parametes associated etcd deployed
 type EtcdSpec struct {
 	// +optional
-	DefragmentationSchedule string `json:"defragmentationSchedule,omitempty"`
-	// +optional
-	ServerPort int `json:"serverPort,omitempty"`
+	Quota *resource.Quantity `json:"quota,omitempty"`
 	// +optional
-	ClientPort int `json:"clientPort,omitempty"`
-	// +required
-	ImageRepository string `json:"imageRepository"`
-	// +required
-	ImageVersion string `json:"imageVersion"`
+	DefragmentationSchedule *string `json:"defragmentationSchedule,omitempty"`
 	// +optional
-	MetricLevel MetricLevel `json:"metrics,omitempty"`
+	ServerPort *int `json:"serverPort,omitempty"`
 	// +optional
-	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+	ClientPort *int `json:"clientPort,omitempty"`
 	// +required
-	EnableTLS bool `json:"enableTLS"`
+	version string `json:"version"`
 	// +optional
-	PullPolicy corev1.PullPolicy `json:"pullPolicy,omitempty"`
+	AuthSecretRef *corev1.SecretReference `json:"authSecretRef, omitempty"`
 	// +optional
-	Username string `json:"username,omitempty"`
+	Metrics MetricsLevel `json:"metrics,omitempty"`
 	// +optional
-	Password string `json:"password,omitempty"`
-	// +required
-	InitialClusterToken string `json:"initialClusterToken"`
```

Comment:
Should we remove InitialCluster details? It may come in handy when we go multi instance. Or do we add it then?

Reply:
Yes. Precisely we are not yet sure about the mutli node setup. We can make if configurable better later than now.